### ISSUE TITLE
Upserting Many-To-Many fails with unique index constraint

### ISSUE THREAD

**Issue by JohnnyCrazy** [prisma/prisma#2094]
## Bug description

Upserting a many-to-many relation fails with a `Unique constraint failed on the fields: (`id`)` error. Some of the executed queries contain a `WHERE 0=1`, so it won't find the counter-part relation and tries to create it. This fails because the `id` key is unique, which is correct.

The complete output below contains some more parameters from my app, they can be ignored.

<details>
<summary>Complete Output</summary>

```
prisma:query BEGIN
2020-04-04T15:09:19.043Z engine stdout {
  timestamp: 'Apr 04 15:09:19.042',
  level: 'INFO',
  target: 'quaint::connector::metrics',
  fields: {
    query: 'SELECT "public"."User"."id" FROM "public"."User" WHERE "public"."User"."id" = $1',
    item_type: 'query',
    params: '["1122095781"]',
    duration_ms: 0
  }
}
prisma:query SELECT "public"."User"."id" FROM "public"."User" WHERE "public"."User"."id" = $1
2020-04-04T15:09:19.044Z engine stdout {
  timestamp: 'Apr 04 15:09:19.042',
  level: 'INFO',
  target: 'quaint::connector::metrics',
  fields: {
    query: 'SELECT "B", "A" FROM "public"."_PlaylistToUser" WHERE "B" IN ($1)',
    item_type: 'query',
    params: '["1122095781"]',
    duration_ms: 0
  }
}
prisma:query SELECT "B", "A" FROM "public"."_PlaylistToUser" WHERE "B" IN ($1)
2020-04-04T15:09:19.045Z engine stdout {
  timestamp: 'Apr 04 15:09:19.043',
  level: 'INFO',
  target: 'quaint::connector::metrics',
  fields: {
    query: 'SELECT "public"."Playlist"."id" FROM "public"."Playlist" WHERE ("public"."Playlist"."id" = $1 AND 1=0) OFFSET $2',
    item_type: 'query',
    params: '["what",0]',
    duration_ms: 0
  }
}
prisma:query SELECT "public"."Playlist"."id" FROM "public"."Playlist" WHERE ("public"."Playlist"."id" = $1 AND 1=0) OFFSET $2
2020-04-04T15:09:19.046Z engine stdout {
  timestamp: 'Apr 04 15:09:19.044',
  level: 'INFO',
  target: 'quaint::connector::metrics',
  fields: {
    query: 'INSERT INTO "public"."Playlist" ("id","name","ownerId","ownerDisplayName","ownerType","public","collaborative","uri","createdAt","updatedAt") VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10) RETURNING "public"."Playlist"."id"',
    item_type: 'query',
    params: '["what","Hello World","johnny","Johnny","user",false,false,"uri",2020-04-04 15:09:19.041 UTC,2020-04-04 15:09:19.041258900 UTC]',
    duration_ms: 1
  }
}
prisma:query INSERT INTO "public"."Playlist" ("id","name","ownerId","ownerDisplayName","ownerType","public","collaborative","uri","createdAt","updatedAt") VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10) RETURNING "public"."Playlist"."id"
2020-04-04T15:09:19.047Z engine stdout {
  timestamp: 'Apr 04 15:09:19.044',
  level: 'INFO',
  target: 'quaint::connector::metrics',
  fields: {
    query: 'ROLLBACK',
    item_type: 'query',
    params: '[]',
    duration_ms: 0
  }
}
prisma:query ROLLBACK
2020-04-04T15:09:19.049Z engine {
  error: PrismaClientKnownRequestError: Unique constraint failed on the fields: (`id`)
      at NodeEngine.graphQLToJSError (/code/node_modules/@prisma/client/runtime/index.js:1:18716)
      at /code/node_modules/@prisma/client/runtime/index.js:1:16338
      at processTicksAndRejections (internal/process/task_queues.js:97:5)
      at Dataloader.loader (/code/node_modules/@prisma/client/runtime/index.js:1:50647) {
    code: 'P2002',
    meta: { target: [Array] }
  }
}
(node:3231) UnhandledPromiseRejectionWarning: Error: 
Invalid `prisma.user.update()` invocation in
/code/src/services/user.ts:33:26

Unique constraint failed on the fields: (`id`)
    at PrismaClientFetcher.request (/code/node_modules/@prisma/client/runtime/index.js:1:51353)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
(node:3231) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:3231) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

</details>

## How to reproduce

Schema:

```
model User {
  id            String      @id
  playlists     Playlist[]  @relation(references: [id])
}

model Playlist {
  id               String   @id
  name             String
  users            User[]   @relation(references: [id])
}
```

1. Add two users, id "1" and "2"
1. Run the following sample for both ids, second will fail:

```
    await prisma.user.update({
      where: { id: user.id },
      data: {
        playlists: {
          upsert: [
            {
              create: {
                id: 'unique',
                name: 'Hello World',
              },
              update: {
                id: 'unique',
                name: 'Hello World',
              },
              where: { id: 'unique' },
            },
          ],
        },
      },
    });
```

## Expected behavior

On the second run, prisma should detect that there is already a `playlist` with `id = 'unique'`. It should update the playlist object and link it in the relation table.

## Prisma information

Above

## Environment & setup

- OS: Linux
- Database: PostgreSQL
- Prisma version: 2.0.0-beta.1
- Node.js version: v13.12.0

**Comment by pantharshit00** [comment]
Firstly `WHERE 0=1` and `WHERE 1=1` constructs are used as on-off switches in dynamic SQL programming to manage multiple conditions so you will find them used by Prisma in all the places.

Now I think this behaviour is completely expected. While performing a nested query the playlist records fetched by the client are the records connected to the user. That is why you see a query to the join table. 

Now since you are managing the ids and they are not being automatically generated, now if you try to insert a record from another record which the playlist record is not connected so it is not being looked up via `where` in the upsert and it will try to insert the record. Since id is unique it will fail as database will reject it.

I will suggest you to not the manage the IDs yourself in order to avoid these situations.

**Comment by JohnnyCrazy** [comment]
I would expect upsert to behave differently and act directly on the relation table, not on the join table. Could you give an example, where a many-to-many upsert workflow like you described works?

> Now since you are managing the ids and they are not being automatically generated, now if you try to insert a record from another record which the playlist record is not connected so it is not being looked up via where in the upsert and it will try to insert the record. Since id is unique it will fail as the database will reject it.
>
> I will suggest you to not to manage the IDs yourself in order to avoid these situations.

But this would completely remove the idea of the upsert. I would like to give the broader example of my use-case:

In Spotify, a user can follow multiple playlists. A playlists can also be followed by multiple playlists. Those playlists have a unique ID assigned by spotify. A new user registered in your application and you need to fetch all of his followed playlists. It makes sense to use upsert here, because maybe another registered user already follows that playlist. If you would use an autogenerated ID, it would create a new one.

**Comment by mavilein** [comment]
I discussed this with @dpetrick and we realized that this behavior is correct. Let's have a look at the nested write in question:
```ts
await prisma.user.update({
      where: { id: user.id },
      data: {
        playlists: {
          upsert: [
            {
              create: {
                id: 'unique',
                name: 'Hello World',
              },
              update: {
                id: 'unique',
                name: 'Hello World',
              },
              where: { id: 'unique' },
            },
          ],
        },
      },
    });
```

The problem is the semantic of this operation. Here is a verbal description of it:
> Update the playlists of a user. Look for related playlists of a user. If a related playlist with id 'unique' exists update this record. Otherwise create a new playlist and connect it to the user.

**The important part is that the playlist record needs be connected to the user.**

Here is an explanation of why that script fails:
1. In the write for user 1 there is no playlist connected to that user yet. Hence playlist with id 'unique' is created and connected to user 1.
2. In the write for user 2 there is no playlist connected to that user yet. The playlist with id 'unique' exists but it's **not** connected to the user. So it's not satisfying the condition and hence we try to create a new record which fails.

TLDR: This behaves as expected. What you probably want is the missing feature for `connectOrCreate` in our API: https://github.com/prisma/prisma-client-js/issues/336

**Comment by JohnnyCrazy** [comment]
Thanks for the detailed explanation @mavilein. Indeed I was looking for the behavior of the linked issue. 👍
